File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-hdinsight-base/src/main/java/com/microsoft/intellij/AzureActionsListener.java
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */

package com.microsoft.intellij;

import com.google.gson.Gson;
import com.intellij.ide.AppLifecycleListener;
import com.microsoft.azure.cosmosspark.serverexplore.cosmossparknode.CosmosSparkClusterOps;
import com.microsoft.azure.hdinsight.common.HDInsightHelperImpl;
import com.microsoft.azure.hdinsight.common.HDInsightLoader;
import com.microsoft.azure.toolkit.intellij.common.CommonConst;
import com.microsoft.azure.toolkit.lib.Azure;
import com.microsoft.azure.toolkit.lib.common.messager.ExceptionNotification;
import com.microsoft.azure.toolkit.lib.common.operation.AzureOperation;
import com.microsoft.azuretools.authmanage.CommonSettings;
import com.microsoft.azuretools.azurecommons.util.FileUtil;
import com.microsoft.azuretools.core.mvp.ui.base.AppSchedulerProvider;
import com.microsoft.azuretools.core.mvp.ui.base.MvpUIHelperFactory;
import com.microsoft.azuretools.core.mvp.ui.base.SchedulerProviderFactory;
import com.microsoft.azuretools.service.ServiceManager;
import com.microsoft.azuretools.telemetry.AppInsightsClient;
import com.microsoft.intellij.helpers.IDEHelperImpl;
import com.microsoft.intellij.helpers.MvpUIHelperImpl;
import com.microsoft.intellij.helpers.UIHelperImpl;
import com.microsoft.intellij.secure.IdeaTrustStrategy;
import com.microsoft.intellij.util.PluginUtil;
import com.microsoft.tooling.msservices.components.DefaultLoader;
import com.microsoft.tooling.msservices.components.PluginComponent;
import com.microsoft.tooling.msservices.components.PluginSettings;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.ssl.TrustStrategy;
import rx.internal.util.PlatformDependent;

import javax.annotation.Nonnull;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.List;
import java.util.logging.FileHandler;

@Slf4j
public class AzureActionsListener implements AppLifecycleListener, PluginComponent {
    public static final String PLUGIN_ID = CommonConst.PLUGIN_ID;
    private static final String AZURE_TOOLS_FOLDER = ".AzureToolsForIntelliJ";
    private static final String AZURE_TOOLS_FOLDER_DEPRECATED = "AzureToolsForIntelliJ";
    private static final FileHandler logFileHandler = null;

    private PluginSettings settings;

    @Override
    @ExceptionNotification
    public void appFrameCreated(@Nonnull List<String> commandLineArgs) {
        try {
            DefaultLoader.setPluginComponent(this);
            DefaultLoader.setUiHelper(new UIHelperImpl());
            DefaultLoader.setIdeHelper(new IDEHelperImpl());
            AppInsightsClient.setAppInsightsConfiguration(new AppInsightsConfigurationImpl());
            SchedulerProviderFactory.getInstance().init(new AppSchedulerProvider());
            MvpUIHelperFactory.getInstance().init(new MvpUIHelperImpl());
            HDInsightLoader.setHHDInsightHelper(new HDInsightHelperImpl());
            try {
                loadPluginSettings();
            } catch (final IOException e) {
                PluginUtil.displayErrorDialogAndLog("Error", "An error occurred while attempting to load settings", e);
            }

            ServiceManager.setServiceProvider(TrustStrategy.class, IdeaTrustStrategy.INSTANCE);
            CommonSettings.setUserAgent(Azure.az().config().getUserAgent());
            initAuthManage();
        } catch (final Throwable t) {
            log.error(t.getMessage(), t);
        }
        try {
            PlatformDependent.isAndroid();
        } catch (final Throwable ignored) {
            DefaultLoader.getUIHelper().showError("A problem with your Android Support plugin setup is preventing the"
                + " Azure Toolkit from functioning correctly (Retrofit2 and RxJava failed to initialize)"
                + ".\nTo fix this issue, try disabling the Android Support plugin or installing the "
                + "Android SDK", "Azure Toolkit for IntelliJ");
            // DefaultLoader.getUIHelper().showException("Android Support Error: isAndroid() throws " + ignored
            //         .getMessage(), ignored, "Error Android", true, false);
        }
    }

    private void initAuthManage() {
        try {
            final String baseFolder = FileUtil.getDirectoryWithinUserHome(AZURE_TOOLS_FOLDER).toString();
            final String deprecatedFolder = FileUtil.getDirectoryWithinUserHome(AZURE_TOOLS_FOLDER_DEPRECATED).toString();
            CommonSettings.setUpEnvironment(baseFolder, deprecatedFolder);
        } catch (final IOException ex) {
            log.error("initAuthManage()", ex);
        }
    }

    @Override
    public PluginSettings getSettings() {
        return settings;
    }

    @Override
    public String getPluginId() {
        return PLUGIN_ID;
    }

    private void loadPluginSettings() throws IOException {
        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(
            AzureActionsListener.class.getResourceAsStream("/settings.json")))) {
            final StringBuilder sb = new StringBuilder();
            String line;

            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }

            final Gson gson = new Gson();
            settings = gson.fromJson(sb.toString(), PluginSettings.class);
        }
    }
}

File: PluginsAndFeatures/azure-toolkit-for-rider/src/main/kotlin/com/microsoft/intellij/AzureActionsListener.kt
package com.microsoft.intellij

import com.intellij.ide.AppLifecycleListener
import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.DefaultActionGroup
import com.intellij.openapi.actionSystem.IdeActions
import com.intellij.util.EnvironmentUtil
import com.microsoft.azure.toolkit.ide.common.store.AzureStoreManager
import com.microsoft.azure.toolkit.ide.common.store.DefaultMachineStore
import com.microsoft.azure.toolkit.intellij.common.action.IntellijAzureActionManager
import com.microsoft.azure.toolkit.intellij.common.auth.IntelliJSecureStore
import com.microsoft.azure.toolkit.intellij.common.messager.IntellijAzureMessager
import com.microsoft.azure.toolkit.intellij.common.settings.IntellijStore
import com.microsoft.azure.toolkit.intellij.common.task.IntellijAzureTaskManager
import com.microsoft.azure.toolkit.lib.common.messager.AzureMessager
import com.microsoft.azure.toolkit.lib.common.task.AzureRxTaskManager
import com.microsoft.azure.toolkit.lib.common.task.AzureTaskManager
import com.microsoft.azure.toolkit.lib.common.utils.CommandUtils
import com.microsoft.azuretools.authmanage.CommonSettings
import com.microsoft.azuretools.azurecommons.util.FileUtil
import com.microsoft.intellij.helpers.IDEHelperImpl
import com.microsoft.intellij.helpers.UiHelperImpl
import com.microsoft.intellij.util.PluginHelper
import com.microsoft.tooling.msservices.components.DefaultLoader
import com.microsoft.tooling.msservices.components.PluginComponent
import com.microsoft.tooling.msservices.components.PluginSettings
import com.microsoft.tooling.msservices.serviceexplorer.Node

class AzureActionsListener: AppLifecycleListener, PluginComponent {
    companion object{
        private const val AZURE_TOOLS_FOLDER = ".AzureToolsForIntelliJ"
        private const val AZURE_TOOLS_FOLDER_DEPRECATED = "AzureToolsForIntelliJ"
    }

    override fun appFrameCreated(commandLineArgs: MutableList<String>) {
        DefaultLoader.setPluginComponent(this)
        DefaultLoader.setUiHelper(UiHelperImpl())
        DefaultLoader.setIdeHelper(IDEHelperImpl())
        AzureTaskManager.register(IntellijAzureTaskManager())
        AzureRxTaskManager.register()
        AzureStoreManager.register(
            DefaultMachineStore(PluginHelper.getTemplateFile("azure.json")),
            IntellijStore.getInstance(),
            IntelliJSecureStore.getInstance()
        )
        AzureInitializer.initialize()
        AzureMessager.setDefaultMessager(IntellijAzureMessager())
        IntellijAzureActionManager.register()
        Node.setNode2Actions(mutableMapOf())
        CommandUtils.setEnv(EnvironmentUtil.getEnvironmentMap())

        initAuthManage()
        val am = ActionManager.getInstance()
        val toolbarGroup = am.getAction(IdeActions.GROUP_MAIN_TOOLBAR) as DefaultActionGroup
        toolbarGroup.addAll(am.getAction("AzureToolbarGroup") as DefaultActionGroup)
        val popupGroup = am.getAction(IdeActions.GROUP_PROJECT_VIEW_POPUP) as DefaultActionGroup
        popupGroup.add(am.getAction("AzurePopupGroup"))
    }

    private fun initAuthManage() {
        val baseFolder = FileUtil.getDirectoryWithinUserHome(AZURE_TOOLS_FOLDER).toString()
        val deprecatedFolder = FileUtil.getDirectoryWithinUserHome(AZURE_TOOLS_FOLDER_DEPRECATED).toString()
        CommonSettings.setUpEnvironment(baseFolder, deprecatedFolder)
    }

    override fun getSettings(): PluginSettings {
        TODO("Not yet implemented")
    }

    override fun getPluginId(): String = "com.intellij.resharper.azure"
}